lib/service.go

```diff
@@ -130,126 +169,173 @@ func (m *RelayService) ProposeBlindedBlockV1(r *http.Request, args *SignedBlinde
 		blockHash = body.ExecutionPayloadCamel.BlockHashCamel
 	}
 
-	payloadCached := m.store.Get(common.HexToHash(blockHash))
+	payloadCached := m.store.GetExecutionPayload(common.HexToHash(blockHash))
 	if payloadCached != nil {
 		logMethod.WithFields(logrus.Fields{
 			"blockHash": payloadCached.BlockHash,
 			"number":    payloadCached.Number,
 			"txRoot":    fmt.Sprintf("%#x", payloadCached.TransactionsRoot),
-		}).Info("ProposeBlindedBlockV1: revealed previous payload from execution client")
+		}).Info("ProposeBlindedBlockV1: revealed previous payload")
 		*result = *payloadCached
 		return nil
 	}
 
-	relayResp, err := makeRequest(m.relayURL, "relay_proposeBlindedBlockV1", []interface{}{args})
-	if err != nil {
-		logMethod.WithFields(logrus.Fields{
-			"error":  err,
-			"url":    m.relayURL,
-			"method": "relay_proposeBlindedBlockV1",
-		}).Error("Could not make request to relay")
-		return err
+	resultC := make(chan *rpcResponseContainer, len(m.relayURLs))
+	for _, url := range m.relayURLs {
+		go func(url string) {
+			res, err := makeRequest(url, "relay_proposeBlindedBlockV1", []interface{}{args})
+			resultC <- &rpcResponseContainer{url, err, res}
+		}(url)
 	}
 
-	resp, err := parseRPCResponse(relayResp)
-	if err != nil {
-		logMethod.WithField("err", err).Error("Could not parse response")
-		return err
-	}
+	for i := 0; i < cap(resultC); i++ {
+		res := <-resultC
 
-	err = json.Unmarshal(resp.Result, result)
-	if err != nil {
-		logMethod.WithField("err", err).Error("Could not unmarshal response")
-		return err
+		// Check for errors
+		if res.err != nil {
+			logMethod.WithFields(logrus.Fields{"error": res.err, "url": res.url}).Error("error making request to relay")
+			continue
+		}
+		if res.res.Error != nil {
+			logMethod.WithFields(logrus.Fields{"error": res.res.Error, "url": res.url}).Warn("error reply from relay")
+			continue
+		}
+
+		// Decode response
+		err := json.Unmarshal(res.res.Result, result)
+		if err != nil {
+			logMethod.WithFields(logrus.Fields{"error": err, "data": string(res.res.Result)}).Error("Could not unmarshal response")
+			continue
+		}
+
+		logMethod.WithFields(logrus.Fields{
+			"blockHash": result.BlockHash,
+			"number":    result.Number,
+			"txRoot":    fmt.Sprintf("%#x", result.TransactionsRoot),
+		}).Info("ProposeBlindedBlockV1: revealed new payload from relay")
+		return nil
```

Comment:
Nit: you could cancel the rest of the requests too. I guess now is not a good time, but it'd be nice to have it at some point.

Reply:
Cool idea! Added a todo and we could revisit this later :)